addons/account/tests/test_account_account.py

```diff
@@ -623,6 +623,71 @@ def test_generate_account_suggestions(self):
         )
         self.assertFalse(account.id in results_2, "Deprecated account should NOT appear in account suggestions")
 
+    def test_placeholder_code(self):
+        """ Test that the placeholder code is '{code_in_company} ({company})'
+            where `company` is the first company in the environment that is
+            in `account.company_ids`.
+
+            Check that `_field_to_sql` gives the same value.
+        """
+        def get_placeholder_code_via_sql(account):
+            account_query = account._as_query()
+            placeholder_code_sql = account_query.select(account._field_to_sql('account_account', 'placeholder_code', account_query))
+            placeholder_code = self.env.execute_query(placeholder_code_sql)[0][0]
+            return placeholder_code
+
+        branch = self.env['res.company'].create({
+            'name': 'Branch',
+            'parent_id': self.company_data['company'].id,
+        })
+        account = self.env['account.account'].create([{
+            'name': 'My account',
+            'company_ids': [Command.set(self.company_data['company'].ids)],
+            'code_mapping_ids': [
+                Command.create({'company_id': self.company_data['company'].id, 'code': '180001'}),
+            ]
+        }])
+
+        # Use `with_company` in order to force an order for `allowed_company_ids` in the context.
+        self.assertEqual(account.with_company(self.company_data['company']).placeholder_code, '180001')
+        self.assertEqual(get_placeholder_code_via_sql(account.with_company(self.company_data['company'])), '180001')
+
+        self.assertEqual(account.with_company(branch).placeholder_code, '180001')
+        self.assertEqual(get_placeholder_code_via_sql(account.with_company(branch)), '180001')
+
+        account_context_1 = account.with_context({
+            'allowed_company_ids': [self.company_data_2['company'].id, self.company_data['company'].id]
+        })
+        self.assertEqual(account_context_1.placeholder_code, '180001 (company_1_data)')
+        self.assertEqual(get_placeholder_code_via_sql(account_context_1), '180001 (company_1_data)')
+
+        account_context_2 = account.with_context({
+            'allowed_company_ids': [self.company_data_2['company'].id, branch.id]
+        })
+        self.assertEqual(account_context_2.placeholder_code, False)
+        self.assertEqual(get_placeholder_code_via_sql(account_context_2), None)
+
+    def test_account_returned_by_search_if_sudo(self):
```

Comment:
Lengthy but clearer (to me :D ).
Feel free to discard if you don't like it.
```suggestion
    def test_account_search_in_sudo_mode_should_bypass_record_rules(self):
```

Reply:
I think that I'll just remove this test as it exists only to validate this bugfix. To me there's not that much value in a test that just makes sure that a LEFT JOIN doesn't get changed back to a normal JOIN :laughing:

Reply:
I won't mind it you remove it but I think you're right and wrong at the same time.
Right because the test seems dumb as it test a basic feature of the ORM.

Wrong because it serves a purpose as a non-regression test.
The fact that it didn't exist anywhere allowed us to break the `search` behavior.
Having it would prevent it to happen again. (And thus, potentially saves some money: ticket created, processed by someone functionally €€, processed by someone else technically €€, maybe fixed by a third person €€ and reviewed by a 4th €€€)

Reply:
I don't know whether `test_account_accessible_by_search_in_sudo_mode` is a good compromise? `test_account_search_in_sudo_mode_should_bypass_the_company_record_rules` seemed a bit too long for me